### ISSUE TITLE
ejabberd: Authentication fails

### ISSUE THREAD

**Issue by mray** [freedombox/Plinth#1058]
In Gajim upon mouseover on contacts in my roster I see these messages for all (8) of my added contacts *except* the one that uses my FreedomBox server as well.

![image](https://user-images.githubusercontent.com/164785/31380519-e71ea0e8-adb1-11e7-9e06-2a90b3148ebf.png)

![image](https://user-images.githubusercontent.com/164785/31380589-06ad1ce6-adb2-11e7-87c9-5903c5bb95bc.png)

![image](https://user-images.githubusercontent.com/164785/31380619-19258174-adb2-11e7-9b50-c5d1da97a1f5.png)

**Comment by mray** [comment]
What can I do about this? I can't communicate with those contacts with this error.

**Comment by JoKeyser** [comment]
@mray is this a recent problem?
My quick searching suggests it might be server-to-server (s2s) settings, see https://www.ejabberd.im/forum/28783/authentication-failed-peer-provided-no-sasl-mechanisms, which links to https://github.com/processone/ejabberd/issues/1700. Maybe you can try their suggestions?

My main question is whether it's your server or theirs that is being complained about. Is your server using a self-signed certificate, or maybe the one from your contacts? You can also check using https://check.messaging.one/

**Comment by mray** [comment]
> @mray is this a recent problem?

Yes, I think it happened after my last update (i'm using the test version of FreedomBox)

**Thanks!** It seems like `mod_s2s_dialback: {}` in the *modules* section of `/etc/ejabberd/ejabberd.yml` was missing. My contacts work fine after adding it and restarting ejabberd via plinth. 

There is a chance this affects more people - we should fix that for everybody.

**Comment by JoKeyser** [comment]
Ah, great that we now know what it is. We still should understand/discuss what this means in terms of privacy, as it seems that this setting enables non-encrypted s2s connections. That should not be a default, but maybe an option (with a warning?):
Citing https://blog.process-one.net/ejabberd-17-03/:
> S2S dialback is now an optional module. If not present in the configuration, only TLS mechanism can be used to establish S2S connections.

**Comment by mray** [comment]
> ... it seems that this setting enables non-encrypted s2s connections.

OH! This is news to me. I'd like to keep s2s encryption enabled if possible!

**Comment by JoKeyser** [comment]
We should understand what it means. I probably formulated it too strongly (didn't have time to read up on it, and probably won't in a few days).
The ultimate resource about the protocol is probably https://xmpp.org/extensions/xep-0220.html. It may "just" be an authentication mechanism, so dropping TLS there doesn't imply that conversations etc are non-encrypted, sorry. Unfortunately, https://docs.ejabberd.im/admin/configuration/ does not talk about dialback. But *somewhere* should be (digestible) information what this option means :).

**Comment by zinid** [comment]
To resolve confusion: there are two authentication mechanisms available for server-to-server connections: [XEP-0178](https://xmpp.org/extensions/xep-0178.html) (aka "SASL EXTERNAL") and [XEP-0220](https://xmpp.org/extensions/xep-0220.html) (aka "dialback").
At the moment the former is the *preferred* one, because dialback is considered less secure (e.g. it is susceptible to DNS poisoning).
In its turn, SASL EXTERNAL means both parties should have TLS certificates signed by trusted CAs (such as Let's Encrypt).

**Comment by zinid** [comment]
However, in practice there is a lot of legacy servers (up to 60%) which either don't have TLS enabled or possess crappy certificates (self-signed, expired, etc). Thus `mod_s2s_dialback` is a MUST if you want to contact those servers.

**Comment by JoKeyser** [comment]
Hi @zinid , thanks a lot for your explanation!
Given this trade-off between security and "usefulness" of the server, I think we should keep the default with dialback switched off, but introduce an option for the admin to allow is. We would then need an easy-to-understand warning what the ramifications are.
@zinid would there be any direct effect on end-to-end verification or encryption? The way I understand it, an attacker could spoof an entire server and thus impersonate any contact, unless a user had verified the contact in question? (And in our use-case, we can't rely on that level of diligence.)

**Comment by zinid** [comment]
> The way I understand it, an attacker could spoof an entire server and thus impersonate any contact

Yes, if an attacker controls the DNS then s/he can generate self-signed certificate and connect with any server with dialback support. So, if you classify DNS attacks as quite possible then dialback is not for you. So this depends on your userbase.

> unless a user had verified the contact in question?

Yes, but this is another story (end-to-end encryption, OMEMO, OTR, etc).

> I think we should keep the default with dialback switched off

Well, if you don't have a valid certificate, then without dialback your s2s connectivity is fully broken. And admins don't like to obtain and configure certificates (not sure about your userbase, though :smile:)

**Comment by JoKeyser** [comment]
> Well, if you don't have a valid certificate, then without dialback your s2s connectivity is fully broken. And admins don't like to obtain and configure certificates (not sure about your userbase, though 😄)

**Comment by mray** [comment]
Illiterate person here. What does that mean for me?

**Comment by zinid** [comment]
@mray for you that means in practice, if you don't focus on security, that dialback+bad_cert is better than no_dialback+good_cert in the terms of connectivity with other servers. In fact the former is the situation you have now: instead of generating a valid certfile you added  `mod_s2s_dialback` and things seem like working fine for you (however there are plenty of servers without dialback, you just don't contact them).

**Comment by mray** [comment]
What does "focus on security" mean? It certainly sounds like something I'm interested in it.

**Comment by JoKeyser** [comment]
@mray I think he refers to the trade-off between security and usefulness, and people differ what is considered "secure enough", i.e. whether to "focus on security" alone, as opposed to usefulness/compatibility with less secure servers. By enabling `mod_s2s_dialback` in your server, it got more useful for you, because your contacts use servers that lack a "good" certificate (where good means that your server would trust automatically). But you traded that usefulness with some security. For people like us in low-risk scenarios, this is might be a better solution than asking your friends to improve or change their server. People in high-risk scenarios, where DNS attacks are likely, that option should be disabled.
So we should choose the default for our most common userbase - I would say in this case that we go along with ejabberd's and Debian's choice to disable per default, but provide an option to enable it (for more usefulness, but less security). My hope is that LE certificates become common enough to overall reduce that number of 60% servers that are less secure.

**Comment by mray** [comment]
Thanks for the explanation. So this is about accepting/rejecting foreign jabber servers with "questionable" certificate quality, where we currently a very high percentage of people use such a low quality certificate server.

So to close the ticket we just need to make sure we ship a ejabber.yaml file with the correct settings?

**Comment by JoKeyser** [comment]
> So to close the ticket we just need to make sure we ship a ejabber.yaml file with the correct settings?
Pretty much - but technically we don't ship the yaml file, but we'd add a checkbox-option to change it through plinth. The defaults already are to use the more secure option.

**Comment by mray** [comment]
Somebody pointed out that my FreedomBox gets a straight `F` on the [IM Observatory](https://check.messaging.one/index.php) - seems to not be such a good deal to lower my standards "a bit" ?

**Comment by JoKeyser** [comment]
Hm, do you get the `F` just because of the enabled dialback? Maybe you could post the (anonymized?) details of this rating?
(I'm getting an `A` for [client-to-server](https://check.messaging.one/result.php?id=75152) and a `B` for [server-to-server](https://check.messaging.one/result.php?id=75153), without remembering any tinkering).

**Comment by zinid** [comment]
`F` usually means the certificate is invalid.

**Comment by mray** [comment]
Yes, it says

> Certificate is not trusted, grade capped to F. Ignoring trust: B.

and

> Error: self signed certificate.

**Comment by zinid** [comment]
Obtain valid certificates, no? :smile:

**Comment by JoKeyser** [comment]
@mray Using an Let's Encrypt certificate for ejabberd should be easy since Plinth 0.16, i.e. if you're on Debian testing. It's the awful UI we're discussing in #1091 :).

**Comment by mray** [comment]
@JoKeyser thanks! I totally forgot to check that option. Can live much better with a `B` :)
